cleverhans/attacks_tf.py

```diff
@@ -133,275 +133,248 @@ def vatm(model, x, logits, eps, num_iterations=1, xi=1e-6,
+        the adversarial samples can fool the model, we do not care about how
+        strong this adversarial samples are.  If you want to finish certain
+        epochs even if we already hava strong adversarial samples, set this to
+        anything larger than 1.0.
+    :param clip_min: The min value for each pixel.
+    :param clip_max: The max value for each pixel.
+
+    :returns: A tensor, the adversarial samples.
     """
-
-    # perturb our input sample
-    if increase:
-        X[0, i] = np.minimum(clip_max, X[0, i] + theta)
-        X[0, j] = np.minimum(clip_max, X[0, j] + theta)
-    else:
-        X[0, i] = np.maximum(clip_min, X[0, i] - theta)
-        X[0, j] = np.maximum(clip_min, X[0, j] - theta)
-
-    return X
-
-
-def saliency_map(grads_target, grads_other, search_domain, increase):
-    """
-    TensorFlow implementation for computing saliency maps
-    :param grads_target: a matrix containing forward derivatives for the
-                         target class
-    :param grads_other: a matrix where every element is the sum of forward
-                        derivatives over all non-target classes at that index
-    :param search_domain: the set of input indices that we are considering
-    :param increase: boolean; true if we are increasing pixels, false otherwise
-    :return: (i, j, search_domain) the two input indices selected and the
-             updated search domain
-    """
-    # Compute the size of the input (the number of features)
-    nf = len(grads_target)
-
-    # Remove the already-used input features from the search space
-    invalid = list(set(range(nf)) - search_domain)
-    increase_coef = (2 * int(increase) - 1)
-    grads_target[invalid] = - increase_coef * np.max(np.abs(grads_target))
-    grads_other[invalid] = increase_coef * np.max(np.abs(grads_other))
-
-    # Create a 2D numpy array of the sum of grads_target and grads_other
-    target_sum = grads_target.reshape((1, nf)) + grads_target.reshape((nf, 1))
-    other_sum = grads_other.reshape((1, nf)) + grads_other.reshape((nf, 1))
-
-    # Create a mask to only keep features that match saliency map conditions
-    if increase:
-        scores_mask = ((target_sum > 0) & (other_sum < 0))
+    xshape = tf.shape(x)
+    n = xshape[0]
+    target = tf.cond(tf.equal(0, tf.rank(y)),
+                     lambda: tf.zeros([n], dtype=tf.int32)+y,
+                     lambda: y)
+
+    if isinstance(epochs, float):
+        tmp = tf.to_float(tf.size(x[0])) * epochs
+        epochs = tf.to_int32(tf.floor(tmp))
+
+    if pair:
+        _jsma_fn = _jsma2_impl
     else:
-        scores_mask = ((target_sum < 0) & (other_sum > 0))
+        _jsma_fn = _jsma_impl
 
-    # Create a 2D numpy array of the scores for each pair of candidate features
-    scores = scores_mask * (-target_sum * other_sum)
+    def _fn(i):
+        # `xi` is of the shape (1, ....), the first dimension is the number of
```

Comment:
This comment is not entirely clear. Could you rephrase a bit? In its current form, it implies that the first dimension is always 1 because the example shape is (1, ...)